examples/spmd/run_spmd.py

```diff
@@ -0,0 +1,188 @@
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        x_t = x.permute(0, 2, 1)
+        return torch.nn.functional.linear(x_t, self.w, self.b)
+
+
+class replicaModel(nn.Module):
+    def __init__(self, layer_count: int = 2, _with_bias: bool = False) -> None:
+        super().__init__()
+        self.seq = nn.Sequential(
+            *[nn.Linear(10, 10, bias=_with_bias) for _ in range(layer_count)]
+        )
+
+    def forward(self, x: torch.Tensor) -> Union[torch.Tensor, Literal[0]]:
+        return sum([self.seq(x)])
+
+
+# ------------ main loop --------------------
+
+
+def work_main(rank: int, world_size: int) -> None:
+    torch.manual_seed(10)
+
+    _device_type = "cuda" if torch.cuda.is_available() else "cpu"
+
+    gpu_placement = torch.arange(
+        world_size
+    )  # .reshape(2, 2) for world_size = 4
+    _debug(f"updated gpu placement = {gpu_placement}")
+
+    mesh = DeviceMesh(device_type=_device_type, mesh=gpu_placement)
+    _debug(f"mesh set to {mesh}\n")
+
+    # control depth of replicaModel
+    layers = 2
+
+    # model = Permute().to(rank)  #
+    model = replicaModel(layer_count=layers).to(_device_type)
+
+    ddp = DDP(deepcopy(model))
+    ddp.to(rank)
+
+    spmd = SPMD(
+        deepcopy(model),
+        schema=Schema(
+            mesh=DeviceMesh(
+                _device_type, gpu_placement  # torch.arange(world_size)
+            ),
+            placements=[Replicate()],
+        ),
+    )
+
+    # model input - need to adjust to match models
+    # permute_input = x = torch.randn(2, 10, 40).to("cuda")
+    x = torch.randn(2, 10).to(_device_type)
+    _debug(f"\ninput tensor, first item = {x[0][0]:.4f}")
+
+    # fire off comms
+    spmd(x).sum().backward()
+    ddp(x).sum().backward()
+
+    if rank == 0:
+        _debug(f" --> backwards run complete, rank {rank}")
+
+        print("Visual of resulting grads:\n")
+        for i, (p1, p2) in enumerate(zip(ddp.parameters(), spmd.parameters())):
+            # just show first row of first 2 grad tensors for quick visual
+            if i < 2:
+                # visual display of initial grads
+                div_grad = p2.grad[0] / world_size
+
+                print(f"DDP:\n {p1.grad[0]}\nSPMD:\n {div_grad}\n")  # type: ignore
+
+            assert p1.grad.allclose(  # type: ignore
+                p2.grad / world_size
+            ), "Mismatch in resulting grads between DDP and SPMD."
+    _debug("--> run completed, all grads matching!")
+
+
+# --------- main above -------------------------
+
+
+def main(rank: int, world_size: int, use_cuda: bool = True) -> None:
+
+    # init
+    setup(rank, world_size, use_cuda)
+
+    _world_size = dist.get_world_size()
+    logging.info(f"--> World size = {_world_size}")
+
+    # main work
+    work_main(rank, world_size)
+
+    # teardown
+    teardown(rank)
+
+
+if __name__ == "__main__":
+    os.environ["MASTER_ADDR"] = "localhost"
+    os.environ["MASTER_PORT"] = "29502"
```

Comment:
Prefer to have a random number > 1024. Otherwise the same program executing twice may get the port being occupied message.

Reply:
good idea.  added random port selection with seed. 

Went with the recommended dynamic range for port selection to be safe:

"Dynamic ports—Ports in the range 49152 to 65535 are not assigned, controlled, or registered. They are used for temporary or private ports. They are also known as private or non-reserved ports. Clients should choose ephemeral port numbers from this range..."